File: 1169-42-2127-maximum-employees-to-be-invited-to-a-meeting/1169-42-2127-maximum-employees-to-be-invited-to-a-meeting.js
var maximumInvitations = function(favorite) {
    const n = favorite.length;
    const deg = Array(n).fill(0);
    favorite.forEach(fav => deg[fav]++);

    const q = [];
    const chain = Array(n).fill(1);

    for (let i = 0; i < n; i++) {
        if (deg[i] === 0) q.push(i);
    }

    while (q.length > 0) {
        const x = q.shift();
        const y = favorite[x];
        chain[y] = Math.max(chain[y], chain[x] + 1);
        deg[y]--;
        if (deg[y] === 0) q.push(y);
    }

    let mx = 0, meetings = 0;
    for (let i = 0; i < n; i++) {
        if (deg[i] === 0) continue;
        let cur = i, length = 0;
        while (deg[cur] > 0) {
            length++;
            deg[cur] = 0;
            cur = favorite[cur];
        }
        if (length > 2) {
            mx = Math.max(mx, length);
        } else {
            meetings += chain[i] + chain[favorite[i]];
        }
    }

    return Math.max(mx, meetings);
};

File: 1169-42-2127-maximum-employees-to-be-invited-to-a-meeting/1169-42-2127-maximum-employees-to-be-invited-to-a-meeting.java
import java.util.*;

class Solution {
    public int maximumInvitations(int[] favorite) {
        int n = favorite.length;
        int[] deg = new int[n];
        for (int fav : favorite) {
            deg[fav]++;
        }

        Queue<Integer> q = new LinkedList<>();
        int[] chain = new int[n];
        Arrays.fill(chain, 1);

        for (int i = 0; i < n; i++) {
            if (deg[i] == 0) {
                q.add(i);
            }
        }

        while (!q.isEmpty()) {
            int x = q.poll();
            int y = favorite[x];
            chain[y] = Math.max(chain[y], chain[x] + 1);
            deg[y]--;
            if (deg[y] == 0) {
                q.add(y);
            }
        }

        int mx = 0, meetings = 0;
        for (int i = 0; i < n; i++) {
            if (deg[i] == 0) {
                continue;
            }
            int cur = i, length = 0;
            while (deg[cur] > 0) {
                length++;
                deg[cur] = 0;
                cur = favorite[cur];
            }
            if (length > 2) {
                mx = Math.max(mx, length);
            } else {
                meetings += chain[i] + chain[favorite[i]];
            }
        }

        return Math.max(mx, meetings);
    }
}
